sample/transforms.go

```diff
@@ -0,0 +1,137 @@
+package sample
+
+import (
+	"cmp"
+	"errors"
+	"math"
+	"slices"
+
+	pq "github.com/emirpasic/gods/v2/queues/priorityqueue"
+	"gonum.org/v1/gonum/floats"
+)
+
+type Transform interface {
+	Apply([]float64) ([]float64, error)
+}
+
+// TODO(parthsareen): potentially cache softmax values
+func softmax(logits []float64) []float64 {
+	var sum float64
+	probs := make([]float64, len(logits))
+	for i, v := range logits {
+		probs[i] = math.Exp(v)
+		sum += probs[i]
+	}
+	floats.Scale(1/sum, probs)
+	return probs
+}
+
+type Temperature float64
+
+func (t Temperature) Apply(logits []float64) ([]float64, error) {
+	if t == 0 {
+		return nil, errors.New("use Greedy sampler instead of Temperature(0)")
+	}
+	if t < 0 || t > 2 {
+		return nil, errors.New("temperature must be between 0 and 2")
+	}
+	temp := math.Max(float64(t), 1e-7)
+
+	// subtracting max logit to avoid under/overflow
+	maxLogit := slices.Max(logits)
+	for i := range logits {
+		logits[i] = (logits[i] - maxLogit) / temp
+	}
+
+	return logits, nil
+}
+
+type logitMap struct {
+	index int
+	logit float64
+}
+
+func logitMapComparator(a, b logitMap) int {
+	return -cmp.Compare(a.logit, b.logit)
+}
+
+type TopK int
+
+// TODO(parthsareen): avoid having to check all logits after this transform
+func (k TopK) Apply(logits []float64) ([]float64, error) {
+	if k <= 0 {
+		return nil, errors.New("k must be greater than 0")
+	}
+	if int(k) >= len(logits) {
+		return logits, nil
+	}
+
+	q := pq.NewWith(logitMapComparator)
```

Comment:
Is the priority queue required here? And if so (but make sure we need it), why not use `container/heap`?

Reply:
cc: @mxyng